### ISSUE TITLE
Proposal handle..missing..property instead of handle && handle.missing && handle.missing.property

### ISSUE THREAD

**Issue by dotnetwise** [babel/babel#2628]
Hi,

This is not yet part of [`ES6`](https://esdiscuss.org/topic/existential-operator-null-propagation-operator#content-65) nor [`ES7`](https://esdiscuss.org/topic/existential-operator-null-propagation-operator#content-65) nor [`TypeScript`](https://github.com/Microsoft/TypeScript/issues/16) agreed specs but I have proposed it already there. So in the meanwhile, just sharing the idea with you guys too.

This would be amazing operator!! Especially for `ES6`/`ES7`/`TypeScript`

```js
var error = a.b.c.d; //this would fail with error if a, b or c are null or undefined.
var current = a && a.b && a.b.c && a.b.c.d; // the current messy way to handle this
var currentBrackets = a && a['b'] && a['b']['c'] && a['b']['c']['d']; //the current messy way to handle this
var typeScript = a?.b?.c?.d; // The typescript way of handling the above mess with no errors
var typeScriptBrackets = a?['b']?['c']?['d']; //The typescript of handling the above mess with no errors
```
However I propose a more clear one - as not to confuse ? from the a ? b : c statements with a?.b statements:

```js
var doubleDots = a..b..c..d; //this would be ideal to understand that you assume that if any of a, b, c is null or undefined the result will be null or undefined.
var doubleDotsWithBrackets = a..['b']..['c']..['d'];
```

For the bracket notation, I recommend two dots instead of a single one as it's consistent with the others when non brackets are used. Hence only the property name is static or dynamic via brackets.

Two dots, means if its null or undefined stop processing further and assume the result of expression is null or undefined. (as d would be null or undefined).

Two dots make it more clear, more visible and more space-wise so you understand what's going on.

This is not messing with numbers too - as is not the same case e.g.

```js
1..toString(); // works returning '1'
var x = {};
x.1 = {y: 'test' }; //fails currently
x[1] = {y: 'test' }; //works currently 
var current = x[1].y; //works
var missing= x[2].y; //throws exception
var assume= x && x[2] && x[2].y; // works but very messy
```

About numbers two options: Your call which one can be adopted, but I recommend first one for compatibility with existing rules!
1. Should fail as it does now (`x.1.y` == `runtime error`)
```js
var err = x..1..y; // should fail as well, since 1 is not a good property name, nor a number to call a method, since it's after x object.
```
2. Should work since it understands that is not a number calling a property from `Number.prototype`
```js
var err = x..1..y; // should work as well, resulting 'test' in this case
var err = x..2..y; // should work as well, resulting undefined in this case
```

With dynamic names:
```js
var correct1 = x..[1]..y; //would work returning 'test'
var correct2 = x..[2]..y; //would work returning undefined;
```

What do you think folks?

**Comment by jmm** [comment]
Hello @dotnetwise, welcome to the Babel community! Nothing like that is going to be added unless it's on a standards track. I think that as of v6 even syntax extensions are now in plugins (though I don't know that there's a full public API for creating them yet. FWIW I do hope something like this ends up in the language...dealing with it currently is one example of syntactic chigurh in JS.

**Comment by hsablonniere** [comment]
@jmm So how does it work? I would definitely want someone to implement a plugin to try it out...

**Comment by hzoo** [comment]
This is a similar request to https://github.com/babel/babel/issues/855 right?

**Comment by jmm** [comment]
@hsablonniere I'm sorry, can you rephrase? How does what work? Do you mean how would a plugin that does this work, or how do the current syntax extension plugins work?

@hzoo Yes, thanks! It's essentially the same (doesn't really matter what the specific syntax is). 

@dotnetwise, I'm going to close this as a duplicate of #855. If this ends up on a standards track I'd say definitely raise a new issue though, as I expect there may be interest in making an "official" plugin for that.

**Comment by dotnetwise** [comment]
It's not going to enter into babel before standard I agree. 
I just wanted to raise your attention in actually contributing to the `ES6` / `ES7` proposals and making the above even better. 

Closing it, I doubt anyone will notice it.

As the title said, it's just a proposal for discussion. Perhaps you should label it accordingly instead of closing it.

**Comment by hsablonniere** [comment]
@jmm Sorry, I meant :

* Should someone work on an implementation before formalizing a proposal? Or is it the other way around?
* Can someone build this as a plugin right now (v6)?
  * If yes, what's the roadmap for the documentation of the public API to do syntax extensions?
  * If no, then... how would we try this?

**Comment by jmm** [comment]
@dotnetwise I understand where you're coming from. But the Babel issue tracker isn't an effective venue for that discussion, that's why I closed the issue. If you have a look at the README you can see how it describes the intended use of the issue tracker:

>  For questions and support please visit the <a href="https://babel-slack.herokuapp.com">Slack community</a> or <a href="http://stackoverflow.com/questions/tagged/babeljs">StackOverflow</a>. The Babel issue tracker is <strong>exclusively</strong> for bug reports and feature requests.

As I mentioned, my understanding is that nothing that isn't on standards track is going to be considered here, so a feature request would be premature at this point.

es-discuss and the Babel slack are better venues for that discussion. I imagine a lot of people would be interested in syntax for that landing in the spec, and I imagine a lot of the people that would read / discuss it if it were here probably participate in those other venues as well. If you have something concrete (not necessarily complete) starting a repo to iterate on a draft proposal and mentioning it in the Babel slack wouldn't be a bad idea. (I assume you're familiar with the format the proposals take, but if not you can find out about the process and format here: https://github.com/tc39/ecma262).

**Comment by jmm** [comment]
@hsablonniere Thanks for clarifying.

> Should someone work on an implementation before formalizing a proposal? Or is it the other way around?

It could go either way. I imagine that prototyping with an implementation would help flush out some of the tricky issues and refine the proposal. I think it'd be hard to do it as vaporware (just a proposal) and get it right.

In any case, maybe Babel can help prototype it, but for it to get into Babel (as an "official" plugin) there'd have to be a proposal first (by which I mean a standards track proposal, not a feature request here, of course).

> Can someone build this as a plugin right now (v6)?

I don't know if there's a legit public API for it. I know that allowing plugins to influence the parsing phase is something **sebmck** has been interested in, but [said wasn't going to happen for v6](https://github.com/babel/babel/issues/2168) (see "What about syntax extensions?"). So I don't know if you could make a proper, distributable plugin for it, but if nothing else I don't see why you couldn't make your own private fork and hack whatever you want into Babylon. Or you cold look at current plugins (by current I mean brand-new for v6) that extend syntax to see how they do it and if it's sufficient to allow prototyping this.

> If yes, what's the roadmap for the documentation of the public API to do syntax extensions?

I don't think there is a roadmap right now (**sebmck** just went through a monumental effort to ship v6, so I don't know that he's been doing / thinking anything re: this recently). If there is, I haven't heard about it.

> If no, then... how would we try this?

I noticed that in the other issue @chocolateboy [mentioned the possibility of being able to implement it via sweet.js](https://github.com/babel/babel/issues/855#issuecomment-75444738), so maybe that's something to look into.

**Comment by hsablonniere** [comment]
@jmm Thanks for the details ;-)
